pkg/externalprovider/provider/installconfig.go

```diff
@@ -0,0 +1,46 @@
+package provider
+
+import (
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/installconfig/aws"
+	icazure "github.com/openshift/installer/pkg/asset/installconfig/azure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// InstallConfigExternalProvider describes the methods required for the installconfig asset.
+type InstallConfigExternalProvider interface {
+	// AddToInstallConfigPlatform adds the current platform to the installconfig.
+	AddToInstallConfigPlatform(p *types.Platform) error
+
+	// ValidateInstallConfig validates the install config.
+	ValidateInstallConfig(
+		Config *types.InstallConfig,
+		File *asset.File,
+		AWS *aws.Metadata,
+		Azure *icazure.Metadata,
+	) error
+
+	// PlatformCredsCheck validates the platform credentials.
+	PlatformCredsCheck(
+		Config *types.InstallConfig,
+		File *asset.File,
+		AWS *aws.Metadata,
+		Azure *icazure.Metadata,
```

Comment:
It would make more sense to just have this signature match the dependencies of the asset being genereated:
https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/platformcredscheck.go#L37

oVirt should never have a function that takes `*aws.Metadata` as a parameter.

Reply:
That is true, but the API is designed to be adopted by all providers, so I added all parameters that are used by any provider at this time. Adding new methods is easier than adding new parameters to existing methods.

Reply:
IMHO my suggestion would make it easier and more ergonomic for this API to be adopted by all providers. Also, I meant this as a general suggestion. The one function I picked here is just an example, but this should be applied to all platform interface functions related to an asset. By mapping the parameters to the asset dependencies you ensure that every platform would have all needed resources. It would then be the responsibility of the provider to extract those resources from the dependencies, e.g. oVirt would have *aws.Metadata available but could ignore it

Reply:
Ideally, we would pass on the install config directly, but if we do that we have a circular dependency because the IC structure is in the same package as the code needing to call the new API.

Reply:
Moving the interface to the types package would resolve the dependency right? IMHO it is better if asset package does not depend on platforms.

Reply:
Ah, I see how the circular dependency is a problem now.  I think there should be a way around the circular dependency by having a common interface. The asset package should not really need to depend on the external providers, and I think that is a benefit of the approach. I think moving the definition of the interface to the asset package could resolve this, but it is a bit too much for me to map it all out now.

Reply:
The main issue is this piece of code:

```go
func init() {
	providerRegistry.Register(ovirt.NewOvirtProvider())
}
```

If we can call this from somewhere that is not in the installconfig flow (e.g. some main.go) then that would solve the whole problem as the registry and the API interface could be split into two packages resolving the circular dependency.  However, I don't know what would be a good place to add this.